ui/tests/unit/services/api-test.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import { module, test } from 'qunit';
+import { setupTest } from 'ember-qunit';
+import sinon from 'sinon';
+
+module('Unit | Service | api', function (hooks) {
+  setupTest(hooks);
+
+  hooks.beforeEach(function () {
+    this.apiService = this.owner.lookup('service:api');
+
+    const authService = this.owner.lookup('service:auth');
+    this.setLastFetch = sinon.spy(authService, 'setLastFetch');
+    this.currentToken = sinon.stub(authService, 'currentToken').value('foobar');
+
+    const namespaceService = this.owner.lookup('service:namespace');
+    this.namespace = sinon.stub(namespaceService, 'path').value('another-ns');
+
+    const controlGroupService = this.owner.lookup('service:control-group');
+    this.wrapInfo = { token: 'ctrl-group', accessor: '84tfdfd5pQ5vOOEMxC2o3Ymt' };
+    this.tokenForUrl = sinon.stub(controlGroupService, 'tokenForUrl').returns(this.wrapInfo);
+    this.deleteControlGroupToken = sinon.spy(controlGroupService, 'deleteControlGroupToken');
+
+    const flashMessageService = this.owner.lookup('service:flash-messages');
+    this.info = sinon.spy(flashMessageService, 'info');
+
+    this.url = '/v1/sys/capabilities-self';
+  });
+
+  test('it should set last fetch time', async function (assert) {
+    await this.apiService.setLastFetch({ url: '/v1/sys/health' });
+    assert.true(this.setLastFetch.notCalled, 'Last fetch is not set for polling url');
+
+    await this.apiService.setLastFetch({ url: '/v1/auth/token/lookup-self' });
+    assert.true(this.setLastFetch.calledOnce, 'Last fetch is set for non polling url');
+  });
+
+  test('it should get control group token', async function (assert) {
+    const context = {
+      url: this.url,
+      init: {
+        method: 'GET',
+        headers: { 'X-Vault-Token': 'root' },
+      },
+    };
+
+    this.tokenForUrl.returns(undefined);
```

Comment:
so this overrides what the stub was initially set to return in the `beforeEach` hook? (`this.wrapInfo`) and returns undefined to confirm that the request isn't "rebuilt" by the control group token logic?

Reply:
Yes that's correct. Basically if there is no control group token the middleware shouldn't do anything. I just initiated the stub with a return value but you can return something different at any time.